tests/mock_db.py

```diff
@@ -0,0 +1,25 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from contextlib import contextmanager
+
+
+class MockDB:
```

Comment:
> Mock objects should be converted to fixtures, so they can be easily imported into the tests. As a simple example:

A benefit of the current implementation is that the mocks both:
* Are configurable where needed (`is_initialized()`, etc)
* Contain common testing logic for the component being tested

With your fixture example, it always returns the same `MockDB()`, even though we'd want to configure it. It could be mutably changed after-the-fact, but having immutability in mocks is nice.

> Even better would be to use monkeypatch to achieve these mocks are doing, for example:

With this monkeypatch example, there would be common monkeypatch logic that is repeated. It's possible to try to move common monkeypatch logic into a function, but that's a parallel to what I have right now - common mock logic is shared in the mock class.

> Test fixtures (e.g. MockDB, MockMail, MockQueryPositionStore, MockSettings, etc.) should go in a conftest module, or at least a fixtures module

The convention I'm following right now is that tests and mocks are placed in the same relative modules as they are in the production code. I'm not sure that grouping into a module by programmatic type is a better solution.
Besides, I've consistently named the files with the mock `mock_$type.py`, and the tests `test_$type.py`.

Reply:
Forgot to comment on this earlier.

Regarding fixtures, there's a separate conversation about them on this PR, and I don't think they make sense for these mock implementations either for the same reasons.

However, you are right that there is boilerplate around asserting function calls (e.g.: how many calls, params, etc).
I don't think that this is worth blocking the PR on, but I'll address this right now anyways. The solution I'll use is to still have the mock class implementation, and to also wrap it with a `unittest.mock` when function assertions are needed. This will allow the removal of the `is_x_method_called` variables.

> I assume what you mean by "production code" is implementation code? 

Yeah, I'm referring to non-test code. Perhaps there's a better term than "production code", though "implementation code" doesn't quite feel like the right term either (tests can have mock _implementations_, so the term "implementation" doesn't only apply to non-tests).

> What's the advantage of having a single "mock_" module for each mock? Actually importing the class you are trying to mock and monkeypatching it is much clearer than having to interpret the meaning from the name of the module and creating a mock class from scratch.

There's less room for surprises and footguns when a mock implementation is created.
For example, imagine testing a component "A" that depends on component "B". You want to test "A" in isolation, so you don't want any production code of "B" to be executed.

If you just throw a mock around "B" and monkeypatch it, you're at risk of forgetting to monkeypatch methods (especially if new ones are added to "B" in the future). Then, your tests of "A" will indirectly call real production methods of "B", which could have unintended side-effects. These side-effects may not be immediately apparent - your tests may still pass until an innocuous change in the future causes an unexpected failure.